### PR TITLE
Card styling bits 2

### PR DESCRIPTION
**Pull request by montehurd**
**Summary of changes**
- Reduced vertical space between cells to 1px.
- Capitalize first character of wikidata descrips in recent/saved.
- Retrieve plain text extracts for cards with search data source.
- Clean summaries now extracted for when the article data store is the data source. i.e. used to get summaries for articles loaded when tapping links.
- Updated number of lines shown for title, description and summary.
- Summaries now usually show 4 good lines of text on landscape.
- Much closer matching of font sizes, padding and linespacing to mock - overlaid they're within a pixel or two.
- Collapsed padding above and below description if no description.
- No placeholder image shown if the article doesn't have a lead image.
- Updated placeholder image.

**Before screenshot**

*enwiki > goat* had quite a few issues related to html-to-nsattributed strings parsing. Note the double ellipsis and large padding above the save icon - this padding was partially from the nsattributed string generated from the html-to-nsattributed strings parser:
goat old

**After screenshots**

No more *goat* summary layout issues - note the padding, linespacing and font updates too:
screen shot 2015-10-10 at 7 29 29 pm

*Related* items fetcher now asks the search api for plain text extracts. Here such an extract is used as the article summary:
screen shot 2015-10-10 at 7 28 21 pm

*Saved Pages and Recent* items don't use the search api. We can now extract very nice summaries directly from data store section html:
screen shot 2015-10-10 at 7 41 47 pm

Example layout of articles without lead images - note the description padding has been collapsed on the article which doesn't have a description:
screen shot 2015-10-10 at 10 40 46 pm

Updated placeholder images:
screen shot 2015-10-12 at 2 28 03 pm
screen shot 2015-10-12 at 2 28 22 pm
screen shot 2015-10-12 at 3 40 31 pm
screen shot 2015-10-13 at 4 48 46 pm

### CODE REVIEW

File: MediaWikiKit/MediaWikiKit/MWKSection.h [hunk 1]
```diff
@@ -159,6 +159,9 @@ extern NSString* const MWKSectionShareSnippetXPath;
  */
 - (BOOL)hasTextData;
 
+- (NSString *)summary;
++ (NSString*)cleanSummary:(NSString*)summary;
```
Comment (coreyfloyd) [inline MediaWikiKit/MediaWikiKit/MWKSection.h]: Shouldn't be in the public API

File: MediaWikiKit/MediaWikiKit/MWKSection.m [hunk 2]
```diff
@@ -238,6 +239,63 @@ - (void)removeAllChildren {
...
+    
+    // Cleanups which need to happen before string is shortened.
+    output = [summary wmf_stringByRecursivelyRemovingParenthesizedContent];
+    output = [output wmf_stringByRemovingBracketedContent];
+    
+    // Now ok to shorten so remaining cleanups are faster.
+    output = [output wmf_safeSubstringToIndex:WMFNumberOfExtractCharacters];
+    
+    // Cleanups safe to do on shortened string.
+    return [[[[[[output stringByReplacingOccurrencesOfString:@"&amp;" withString:@"&"]
```
Comment (coreyfloyd) [inline MediaWikiKit/MediaWikiKit/MWKSection.m]: Logic on lines 291, 292, and 293 should be in NSString category method(s).

File: MediaWikiKit/MediaWikiKit/MWKArticle.m [hunk 3]
```diff
@@ -474,119 +475,20 @@ - (NSArray*)citations {
...
-        NSArray* paragraphsWrappedWithParagraphTags =
-            [nonEmptyParagraphsWithSelectedChildren bk_map:^NSString*(NSString* thisParagraph) {
-            return [@[@"<p>", thisParagraph, @"</p>"] componentsJoinedByString : @""];
-        }];
-
-        NSData* xpathData = [[paragraphsWrappedWithParagraphTags componentsJoinedByString:@""] dataUsingEncoding:NSUTF8StringEncoding];
-
-        return [[NSAttributedString alloc] initWithHTMLData:xpathData site:self.site];
     }
+    _summary = nil;
```
Comment (coreyfloyd) [inline MediaWikiKit/MediaWikiKit/MWKArticle.m]: why is this here?

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m [hunk 4]
```diff
@@ -24,6 +30,18 @@ - (void)prepareForReuse {
     self.summaryLabel.text = nil;
 }
 
+-(void)awakeFromNib {
+    [self rememberSettingsFromIB];
+}
+
+-(void)rememberSettingsFromIB {
+    NSRange r = NSMakeRange(0, 1);
+    self.summaryLabelAttributesFromIB = [self.summaryLabel.attributedText attributesAtIndex:0 effectiveRange:&r];
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: We should just do this in code. It's a bit tacky to configure this in IB and then grab it. If you really want to see what it looks like in IB, then we should implement the IBDesignable methods to make that happen.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m [hunk 5]
```diff
@@ -24,6 +30,18 @@ - (void)prepareForReuse {
...
+    self.paddingAboveDescriptionFromIB = self.paddingConstraintAboveDescription.constant;
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Instead of caching the content, did you try setting deactivating the constraint when you want to ignore it?

```
self.paddingConstraintAboveDescription.active = NO
```

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m [hunk 6]
```diff
@@ -24,6 +30,18 @@ - (void)prepareForReuse {
...
+    self.paddingBelowDescriptionFromIB = self.paddingConstraintBelowDescription.constant;
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Same for this one

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 5 above)
Comment (bgerstle) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: you could also do all of it in masonry, and write visual tests to verify the appearance of each state instead of relying on IB... just sayin' :trollface:

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m [hunk 7]
```diff
@@ -24,6 +30,18 @@ - (void)prepareForReuse {
...
+    self.heightOfImageFromIB = self.imageHeightConstraint.constant;
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: For this one, can you just set the vertical content hugging priority to be greater than the priority of the imageHeightConstraint? I would think it would just collapse to 0 in this case.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 5 above)
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: I'd lay off the test suggestions here. We are still running long on this ticket and just need to get it out the door. We can always do visual tests during our beta fix period.

**Comment by bgerstle** [comment]
@montehurd this looks _really_ nice! aside from some minor nits, my only ask is that you create "expected output" tests for the other visual tests that you deleted. you already did it for the obama article, but we should at least have one w/ the Taj Mahal article as well.

**Comment by bgerstle** [comment]
My only other question is: should we always show the `entityDescription` as first-letter-capitalized, and if so, how to manage that while preserving its raw form? i don't like putting the capitalization logic in multiple places, so I would vote for lazily computing & storing the result on `MWKArticle` while reading and writing the "raw" entity description from mobileview & to the data store respectively.

File: MediaWikiKit/MediaWikiKit/MWKArticle.m (hunk 3 above)
Comment (montehurd) [inline MediaWikiKit/MediaWikiKit/MWKArticle.m]: ah good catch!

File: MediaWikiKit/MediaWikiKit/MWKSection.h (hunk 1 above)
Comment (montehurd) [inline MediaWikiKit/MediaWikiKit/MWKSection.h]: fixed

File: MediaWikiKit/MediaWikiKit/MWKSection.m (hunk 2 above)
Comment (montehurd) [inline MediaWikiKit/MediaWikiKit/MWKSection.m]: done!

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 5 above)
Comment (montehurd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: @coreyfloyd active = no isn't what we want i think - we still want this constrain - we just don't want the padding.

**Comment by bgerstle** [comment]
@montehurd not sure if you're already fixing this, but please change the placeholder in `WMFImageCollectionViewCell` so the article view also has the correct placeholder:

image

**Comment by montehurd** [comment]
@bgerstle just fixed the lead image placeholder.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 5 above)
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: @montehurd What is the constraint doing without the padding? I would think nothing correct?

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 7 above)
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: @montehurd Any movement on this one as well?

File: Wikipedia/UI-V5/WMFSaveableTitleCollectionViewCell.m [hunk 8]
```diff
@@ -32,13 +41,13 @@ - (void)prepareForReuse {
...
     [self.imageView wmf_reset];
-    self.imageView.image = [UIImage imageNamed:[[self class] defaultImageName]];
+    [self configureImageViewWithPlaceholder];
 }
 
 #pragma mark - Accessors
 
-+ (NSString*)defaultImageName {
-    return @"lead-default";
+-(UIImage*)placeholderImage {
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFSaveableTitleCollectionViewCell.m]: Is this different per instance? If not it probably should still be a class method.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 4 above)
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Also this wasn't changed either - I think @bgerstle and I are in agreement here that we shouldn't be using IB to get the settings for Attributed Strings. If you want to discus we can, but as I mentioned above its a bit of a hack and I think as Brian as mentioned, it "hides" the configuration of the attributes where no one expects to look.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m [hunk 9]
```diff
@@ -40,38 +59,55 @@ - (UICollectionViewLayoutAttributes*)preferredLayoutAttributesFittingAttributes:
 - (void)setDescriptionText:(NSString*)descriptionText {
     _descriptionText           = descriptionText;
     self.descriptionLabel.text = descriptionText;
+    if (!self.descriptionLabel.text || self.descriptionLabel.text.length == 0) {
```
Comment (coreyfloyd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Only the second check is necessary

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 5 above)
Comment (montehurd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: @coreyfloyd it's still constrains as configured, just without the padding.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 4 above)
Comment (montehurd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Fixed!

File: Wikipedia/UI-V5/WMFSaveableTitleCollectionViewCell.m (hunk 8 above)
Comment (montehurd) [inline Wikipedia/UI-V5/WMFSaveableTitleCollectionViewCell.m]: Fixed.

File: Wikipedia/UI-V5/WMFArticlePreviewCell.m (hunk 9 above)
Comment (montehurd) [inline Wikipedia/UI-V5/WMFArticlePreviewCell.m]: Fixed!
